R/fit_distributions.R

```diff
@@ -1,70 +1,114 @@
+#' @param metric a subset of `mle`, `jaccard`, `intersection`, `ks`, `mse`, `chisq`
+#' indicating metrics to use for fit optimization
+#' @param truncated logical, whether to fit truncated distributions
+#' @param distributions character vector indicating distributions,
+#' subset of `norm`, `gamma`, `gamma_flip` and `unif`.
 #'
-#' @param x numeric vector representing the density of a histogram
-#' @param metric one of `jaccard`, `intersection`, `ks`, `mse`, `chisq`
+#' @export
 #'
-#' @return a list with the following data
+#' @return a nested list where each sublist represents a model with the following data
 #' \describe{
 #'     \item{par}{a character string denoting the region_id of the Histogram}
 #'     \item{dist}{the distribution name}
 #'     \item{metric}{the metric used to fit the distribution}
 #'     \item{value}{the fitted value of the metric function}
 #'     \item{dens}{a function that returns the density of the fitted distribution}
 #' }
-#' @exportS3Method fit_uniform numeric
-fit_uniform.numeric <- function(x, metric=c('mle', 'jaccard', 'intersection', 'ks', 'mse', 'chisq')) {
-  # Error checking
-  metric <- match.arg(metric)
+#'
+#' @importFrom DEoptim DEoptim
+fit_distributions <- function(
+    x,
+    metric = c("jaccard", "intersection", "ks", "mse", "chisq"),
+    truncated = FALSE,
+    distributions = c("norm", "gamma", "gamma_flip", "unif")
+) {
+  UseMethod("fit_distributions")
+}
+
+#' @exportS3Method fit_distributions numeric
+fit_distributions.numeric <- function(
+    x,
+    metric = c("jaccard", "intersection", "ks", "mse", "chisq"),
+    truncated = FALSE,
+    distributions = c("norm", "gamma", "gamma_flip", "unif")
+) {
 
+  # Assume a bin width of 1 for a numeric vector
   L <- length(x)
-  N <- sum(x)
-  bin <- 1:L
-  p_unif <- generate_uniform_distribution(x)
-
-  if(metric == 'mle') {
-    # Negative log-likelihood
-    value <- (- uniform_mle(x, a = 0, b = L, log = TRUE))
-  } else {
-    metric_func <- get(paste('histogram', metric, sep = "."))
-    m <- metric_func(x, p_unif*N)
-    value <- correct_fitted_value(metric, m)
-  }
+  interval_midpoint <- seq(1, L, 1)
+  fit_distributions_helper(
+    x = x,
+    interval_start = interval_midpoint - 0.5,
+    interval_end = interval_midpoint + 0.5,
+    interval_midpoint = interval_midpoint,
+    metric = metric,
+    truncated = truncated,
+    distributions = distributions
+  )
 
-  return(
-    list(
-      "par" = NULL,
-      "dist" = "unif",
-      "metric" = metric,
-      "value" = value,
-      "dens" = function(x = NULL, mpar = NULL, scale = TRUE) {
-        if(missing(x)) {
-          x <- bin
-        }
-        res <- ifelse(x >= min(bin) & x <= max(bin), p_unif[1], 0)
-        if(scale) res * N
-        else res
-      }
-    )
+}
+
+#' @exportS3Method fit_distributions table
+fit_distributions.table <- fit_distributions.numeric
+
+#' @exportS3Method fit_distributions GenomicHistogram
+fit_distributions.GenomicHistogram <- function(
```

Comment:
Use `@rdname fit_distributions` to link documentation as described in comment below.